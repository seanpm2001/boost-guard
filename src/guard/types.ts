export const domain = {
  name: 'boost',
  version: '1',
  chainId: 4,
  verifyingContract: '0x0000000000000000000000000000000000000000'
};

export const claimTypes = {
  Claim: [
    { name: 'boostId', type: 'uint256' },
    { name: 'recipient', type: 'address' },
    { name: 'amount', type: 'uint256' }
  ]
};

export interface Claim {
  boostId: string;
  recipient: string;
  amount: number;
}

export interface Strategy {
  name: string;
  tag: string;
  params: Record<string, any>;
}
