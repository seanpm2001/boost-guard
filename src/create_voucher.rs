use crate::{ServerError, HUB_URL};
use ::axum::extract::Json;
use axum::response::IntoResponse;
use graphql_client::{GraphQLQuery, Response as GraphQLResponse};
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Default, Deserialize, Serialize)]
pub struct CreateVoucherResponse {
    // TODO: should we include ID of request?
    pub signature: String,
    pub boost_id: String,
    pub user: String,
    pub proposal_id: String,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct CreateVoucherParams {
    pub proposal_id: String,
    pub voter_address: String,
    pub boosts: Vec<(String, String)>,
}

#[derive(GraphQLQuery)]
#[graphql(
    schema_path = "src/graphql/schema.graphql",
    query_path = "src/graphql/proposal_query.graphql",
    response_derives = "Debug"
)]
struct ProposalQuery;

// Receives proposal_id, voter_address, and boost_id
// Queries graph to get boost info?
// Have to check proposal's type: needs to be single-choice or basic, else error -> DONE
// Have to check
// Boost info needed: eligiblity_criteria (incentive or bribe?)
pub async fn create_voucher_handler(
    Json(p): Json<Value>,
) -> Result<impl IntoResponse, ServerError> {
    let requests: Vec<CreateVoucherParams> = serde_json::from_value(p).expect("params");

    let client = reqwest::Client::new();
    for request in requests {
        check_proposal_type(client.clone(), &request.proposal_id).await?;
    }

    // Query the hub to get info about the user's vote
    let response = CreateVoucherResponse::default();
    Ok(Json(response))
}

async fn check_proposal_type(
    client: reqwest::Client,
    proposal_id: &str,
) -> Result<(), ServerError> {
    let variables = proposal_query::Variables {
        id: proposal_id.to_owned(),
    };

    let request_body = ProposalQuery::build_query(variables);

    let res = client.post(HUB_URL).json(&request_body).send().await?;
    let response_body: GraphQLResponse<proposal_query::ResponseData> = res.json().await?;
    let proposal: proposal_query::ProposalQueryProposal = response_body
        .data
        .ok_or("missing data from the hub")?
        .proposal
        .ok_or("missing proposal data from the hub")?;

    let proposal_type = proposal.type_.ok_or("missing proposal type from the hub")?;

    if (proposal_type != "single-choice") && (proposal_type != "basic") {
        return Err(ServerError::ErrorString(format!(
            "`{proposal_type:}` proposals are not eligible for boosting"
        )));
    }

    Ok(())
}
