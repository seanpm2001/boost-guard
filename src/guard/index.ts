import snapshot from '@snapshot-labs/snapshot.js';
import { Wallet } from '@ethersproject/wallet';
import { domain, Claim, claimTypes, Strategy } from './types';
import * as strategies from '../strategies';

const BOOST_SUBGRAPH = 'https://api.thegraph.com/subgraphs/name/mktcode/boost';

export class Guard {
  private readonly wallet: Wallet;
  public address: string;

  constructor(wallet: Wallet) {
    this.wallet = wallet;
    this.address = wallet.publicKey;
  }

  public async getMetadata(boostId): Promise<Strategy> {
    const result = await snapshot.utils.subgraphRequest(BOOST_SUBGRAPH, {
      boost: {
        __args: {
          id: boostId
        },
        strategyURI: true,
        tag: true
      }
    });
    const strategyURI = !result.boost.strategyURI.includes('://')
      ? `ipfs://${result.boost.strategyURI}`
      : result.boost.strategyURI;
    return await snapshot.utils.getJSON(snapshot.utils.getUrl(strategyURI));
  }

  public async run(boostId: string, recipient): Promise<Claim> {
    const metadata = await this.getMetadata(boostId);
    const strategy = new strategies[metadata.name].default(
      boostId,
      recipient,
      metadata.tag,
      metadata.params
    );
    return await strategy.run();
  }

  public async claim(message: Claim) {
    const data: any = { domain, types: claimTypes, message };
    const sig = await this.wallet._signTypedData(domain, claimTypes, message);
    return { address: this.address, sig, data };
  }
}
