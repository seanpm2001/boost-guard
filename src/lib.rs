use axum::response::{IntoResponse, Response};
use hyper::http::StatusCode;

pub mod routes;
pub mod signatures;

use std::env;
extern crate dotenv;
use dotenv::dotenv;

#[macro_use]
extern crate lazy_static;

lazy_static! {
    static ref HUB_URL: String = {
        dotenv().ok();
        env::var("HUB_URL").expect("Please add HUB_URL to your environment or .env file.")
    };
    static ref SUBGRAPH_URL: String = env::var("SUBGRAPH_URL")
        .expect("Please add SUBGRAPH_URL to your environment or .env file.");
    static ref BOOST_NAME: String =
        env::var("BOOST_NAME").expect("Please add BOOST_NAME to your environment or .env file");
    static ref BOOST_VERSION: String = env::var("BOOST_VERSION")
        .expect("Please add BOOST_VERSION to your environment or .env file");
    static ref VERIFYING_CONTRACT: String = env::var("VERIFYING_CONTRACT")
        .expect("Please add VERIFYING_CONTRACT to your environment or .env file");
}

#[derive(Debug, PartialEq)]
pub enum ServerError {
    ErrorString(String),
}

impl<T: std::string::ToString + Sized> From<T> for ServerError {
    fn from(err: T) -> Self {
        ServerError::ErrorString(err.to_string())
    }
}

impl IntoResponse for ServerError {
    fn into_response(self) -> Response {
        let ServerError::ErrorString(body) = self;

        (StatusCode::INTERNAL_SERVER_ERROR, body).into_response()
    }
}

#[derive(Debug, Clone)]
pub struct State {
    pub client: reqwest::Client,
    pub wallet: ethers::signers::LocalWallet,
}
