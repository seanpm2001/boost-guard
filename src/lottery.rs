use crate::routes::{every_vote_query, BoostInfo, EveryVoteQuery, ProposalInfo, VoteInfo};
use crate::ServerError;
use crate::HUB_URL;
use cached::proc_macro::cached;
use cached::TimedSizedCache;
use durations::WEEK;
use ethers::types::{Address, U256};
use graphql_client::{GraphQLQuery, Response as GraphQLResponse};
use rand::prelude::*;
use rand_chacha::ChaCha20Rng;
use std::collections::HashMap;
use std::str::FromStr;

// LRU cache that uses `boost_id` and `chain_id` as keys
#[cached(
    result = true,
    sync_writes = true,
    type = "TimedSizedCache<String, HashMap<Address, U256>>",
    create = "{ TimedSizedCache::with_size_and_lifespan(500, 3 * WEEK.as_secs())}",
    convert = r#"{ format!("{}{}", boost_info.id, boost_info.chain_id) }"#
)]
pub async fn cached_lottery_winners(
    client: Option<&reqwest::Client>,
    boost_info: &BoostInfo,
    proposal_info: &ProposalInfo,
    num_winners: u32,
    limit: Option<u16>,
) -> Result<HashMap<Address, U256>, ServerError> {
    let variables = every_vote_query::Variables {
        proposal: proposal_info.id.to_owned(),
    };
    let request_body = EveryVoteQuery::build_query(variables);
    let query_results: GraphQLResponse<every_vote_query::ResponseData> = client
        .expect("client should be here")
        .post(HUB_URL.as_str())
        .json(&request_body)
        .send()
        .await?
        .json()
        .await?;

    let mut votes = query_results
        .data
        .ok_or("lottery: missing data from the hub")?
        .votes
        .ok_or("lottery: missing votes from the hub")?
        .into_iter()
        .map(|v| {
            let vote_data = v.ok_or("lottery: missing vote info from the hub")?;
            let vote = VoteInfo {
                voter: Address::from_str(&vote_data.voter)
                    .map_err(|e| format!("lottery: {:?}", e))?,
                voting_power: vote_data.vp.ok_or("missing vp from the hub")?,
                choice: vote_data.choice,
            };
            Ok::<VoteInfo, ServerError>(vote)
        })
        .collect::<Result<Vec<VoteInfo>, _>>()?;

    if let Some(boosted_choice) = boost_info.params.eligibility.boosted_choice() {
        votes.retain(|v| v.choice == boosted_choice);
    }

    // Every voter is eligible to the same reward!
    if votes.len() <= num_winners as usize {
        let prize = boost_info.pool_size / votes.len() as u32;
        return Ok(votes.into_iter().map(|v| (v.voter, prize)).collect());
    }

    if let Some(limit) = limit {
        adjust_vote_weights(&mut votes, boost_info.decimals, proposal_info.score, limit)?;
    }

    let prize = boost_info.pool_size / num_winners;
    let seed = ChaCha20Rng::from_entropy().gen(); // todo: e.g from block ranDAO reveal

    Ok(draw_winners(votes, seed, num_winners, prize))
}

// Adjust the voting power of the voters to respect the limit.
// The limit is given in base `10_000`, meaning a limit of `1000` means "no one should have more than 10% chances of getting picked".
// To enforce that, we iterate through the voters' voting power and adjust their voting power to respect the limit.
// If there are not enough voters to reach the limit, the limit will be ignored (e.g: if the limit is 1%, and there is only one voter,
// the voter will get 100% of the prize). Limit will be ignored if set to 0.
// The array of `votes` is assumed to be sorted by voting power.
fn adjust_vote_weights(
    votes: &mut [VoteInfo],
    decimals: u8,
    score: f64,
    limit: u16,
) -> Result<(), ServerError> {
    // Ensure the vector is sorted
    if votes
        .windows(2)
        .any(|w| w[0].voting_power < w[1].voting_power)
    {
        return Err(ServerError::ErrorString("votes are not sorted".to_string()));
    }

    if limit == 0 {
        // log needed
        return Ok(());
    }

    if votes.len() < (10_000.0 / limit as f64).ceil() as usize {
        // log needed "not enough voters to enforce the limit"
        return Ok(());
    }

    let pow = 10_f64.powi(decimals as i32);

    // The "voting power" remaining. At each iteration, we will subtract the voting power of the voter.
    let mut remaining_score = U256::from((score * pow) as u128);

    // The "adjust voting power" remaining. At each iteration, we will subtract the adjusted voting power of the voter.
    let mut adjusted_remaining_score = remaining_score;

    // The maximum adjusted voting power that can be assigned to any voter
    let vp_limit = remaining_score * limit / 10_000; // TODO: constant

    votes.iter_mut().for_each(|v| {
        let vp = U256::from((v.voting_power * pow) as u128);
        // If the user reaches the limit, assign the limit, else assign the correct ratio.
        let adjusted_voting_power =
            std::cmp::min(vp_limit, adjusted_remaining_score * vp / remaining_score);

        // Subtract the voting power
        remaining_score -= vp;
        // Subtract the adjusted voting power
        adjusted_remaining_score -= adjusted_voting_power;

        // Update the voter's voting power
        v.voting_power = adjusted_voting_power.as_u128() as f64 / pow;
    });

    Ok(())
}

fn draw_winners(
    votes: Vec<VoteInfo>,
    seed: u64,
    num_winners: u32,
    prize: U256,
) -> HashMap<Address, U256> {
    let mut rng = ChaCha20Rng::seed_from_u64(seed);
    let mut winners = HashMap::with_capacity(num_winners as usize);

    let mut set = std::collections::HashSet::new();

    // Construct the cumulative weights (e.g; [1, 2, 3, 4] -> [1, 3, 6, 10])
    let mut cumulative_weights = Vec::with_capacity(votes.len());
    let mut curr = votes[0].voting_power;
    cumulative_weights.push(curr);
    for v in votes.iter().skip(1) {
        curr += v.voting_power;
        cumulative_weights.push(curr);
    }

    // TODO: we could optimize by sorting by votes and then poping the last element (which has the highest
    // probability of getting picked). For now, we don't optimize.
    let range = 0.0..*cumulative_weights.last().unwrap();
    for _ in 0..num_winners {
        let winner = loop {
            // Generate a random number between 0 and the highest element of the cumulative weights
            let rnd: f64 = rng.gen_range(range.clone());
            // Get the index of the first element that is greater than or equal to the random number
            let idx = cumulative_weights.iter().position(|x| *x >= rnd).unwrap();
            // Get the corresponding winner address
            let winner = votes.get(idx).unwrap().voter;

            // If the winner has been selected before, draw again.
            if set.contains(&winner) {
                continue;
            } else {
                break winner;
            }
        };

        // Add winner to the set
        set.insert(winner);
        // Add winner to the winners map
        winners.insert(winner, prize);
    }
    winners
}

#[cfg(test)]
mod test_draw_winners {
    use super::draw_winners;
    use super::VoteInfo;
    use super::U256;
    use rand::Rng;
    use rand::SeedableRng;
    use rand_chacha::ChaCha8Rng;

    #[test]
    fn test_randomness() {
        let vote1 = VoteInfo {
            voting_power: 99.0,
            ..Default::default()
        };
        let vote2 = VoteInfo {
            voting_power: 1.0,
            ..Default::default()
        };
        let votes = vec![vote1.clone(), vote2.clone()];
        println!("{}, {}", vote1.voter, vote2.voter);
        let prize = U256::from(10);

        let mut rng = ChaCha8Rng::from_entropy();
        let mut num = 0;

        // Draw 10000 times, expect voter 2 to get picked about 100 times.
        for _ in 0..10000 {
            let winners = draw_winners(votes.clone(), rng.gen(), 1, prize);
            if winners.get(&vote2.voter).is_some() {
                num += 1;
            }
        }

        // Allow for a margin of error
        assert!(num >= 70);
        assert!(num <= 130);
    }

    #[test]
    fn select_two() {
        let vote1 = VoteInfo {
            voting_power: 98.0,
            ..Default::default()
        };
        let vote2 = VoteInfo {
            voting_power: 1.0,
            ..Default::default()
        };
        let vote3 = VoteInfo {
            voting_power: 1.0,
            ..Default::default()
        };
        let votes = vec![vote1.clone(), vote2.clone(), vote3.clone()];
        println!("{}, {}, {}", vote1.voter, vote2.voter, vote3.voter);
        let prize = U256::from(10);

        let mut rng = ChaCha8Rng::from_entropy();

        let winners = draw_winners(votes, rng.gen(), 2, prize);
        assert_eq!(winners.len(), 2);
    }

    #[test]
    fn test_speed() {
        let votes = (0..1000000)
            .enumerate()
            .map(|(i, _)| VoteInfo {
                voting_power: i as f64,
                ..Default::default()
            })
            .collect();
        let prize = U256::from(10);

        let mut rng = ChaCha8Rng::from_entropy();

        let start = std::time::Instant::now();
        println!("start");
        let _ = draw_winners(votes, rng.gen(), 1000, prize);
        let finish = std::time::Instant::now();
        println!("Time: {:?}", finish - start);
    }
}

#[cfg(test)]
mod test_adjust_vote_weights {
    use super::adjust_vote_weights;
    use super::VoteInfo;

    #[test]
    fn test_adjust_vote_weights_half() {
        let mut votes = vec![
            VoteInfo {
                voting_power: 900.0,
                ..Default::default()
            },
            VoteInfo {
                voting_power: 100.0,
                ..Default::default()
            },
        ];
        let decimals = 18;
        let score = votes.iter().map(|v| v.voting_power).sum::<f64>();
        let limit = 5000; // 50 %

        adjust_vote_weights(&mut votes, decimals, score, limit).unwrap();

        assert_eq!(votes[0].voting_power, 500.0);
        assert_eq!(votes[1].voting_power, 500.0);
    }

    #[test]
    fn test_adjust_no_op() {
        let mut votes = vec![
            VoteInfo {
                voting_power: 900.0,
                ..Default::default()
            },
            VoteInfo {
                voting_power: 100.0,
                ..Default::default()
            },
        ];
        let decimals = 18;
        let score = votes.iter().map(|v| v.voting_power).sum::<f64>();
        let limit = 100; // 1 %

        adjust_vote_weights(&mut votes, decimals, score, limit).unwrap();

        assert_eq!(votes[0].voting_power, 900.0);
        assert_eq!(votes[1].voting_power, 100.0);
    }

    #[test]
    fn test_adjust_limit_zero() {
        let mut votes = vec![
            VoteInfo {
                voting_power: 900.0,
                ..Default::default()
            },
            VoteInfo {
                voting_power: 100.0,
                ..Default::default()
            },
        ];
        let decimals = 18;
        let score = votes.iter().map(|v| v.voting_power).sum::<f64>();
        let limit = 0; // 0 %

        adjust_vote_weights(&mut votes, decimals, score, limit).unwrap();

        assert_eq!(votes[0].voting_power, 900.0);
        assert_eq!(votes[1].voting_power, 100.0);
    }

    #[test]
    fn test_adjust_limit_no_op_rounded() {
        let mut votes = vec![
            VoteInfo {
                voting_power: 900.0,
                ..Default::default()
            },
            VoteInfo {
                voting_power: 50.0,
                ..Default::default()
            },
            VoteInfo {
                voting_power: 50.0,
                ..Default::default()
            },
        ];
        let decimals = 18;
        let score = votes.iter().map(|v| v.voting_power).sum::<f64>();
        let limit = 3000; // 30 %

        // Would need 4 voters but we only have three so no-op
        adjust_vote_weights(&mut votes, decimals, score, limit).unwrap();

        assert_eq!(votes[0].voting_power, 900.0);
        assert_eq!(votes[1].voting_power, 50.0);
        assert_eq!(votes[2].voting_power, 50.0);
    }

    #[test]
    fn test_adjust_limit_rounded() {
        let mut votes = vec![
            VoteInfo {
                voting_power: 800.0,
                ..Default::default()
            },
            VoteInfo {
                voting_power: 100.0,
                ..Default::default()
            },
            VoteInfo {
                voting_power: 50.0,
                ..Default::default()
            },
            VoteInfo {
                voting_power: 50.0,
                ..Default::default()
            },
        ];
        let decimals = 18;
        let score = votes.iter().map(|v| v.voting_power).sum::<f64>();
        let limit = 3000; // 30 %

        // We indeed have 4 voters, votes should get adjusted
        adjust_vote_weights(&mut votes, decimals, score, limit).unwrap();

        assert_eq!(votes[0].voting_power, 300.0);
        assert_eq!(votes[1].voting_power, 300.0);
        assert_eq!(votes[2].voting_power, 200.0);
        assert_eq!(votes[2].voting_power, 200.0);
    }

    #[test]
    fn test_adjust_vote_weights() {
        let mut votes = vec![
            VoteInfo {
                voting_power: 458.0,
                ..Default::default()
            },
            VoteInfo {
                voting_power: 200.0,
                ..Default::default()
            },
            VoteInfo {
                voting_power: 180.0,
                ..Default::default()
            },
            VoteInfo {
                voting_power: 150.0,
                ..Default::default()
            },
            VoteInfo {
                voting_power: 5.0,
                ..Default::default()
            },
            VoteInfo {
                voting_power: 4.0,
                ..Default::default()
            },
            VoteInfo {
                voting_power: 3.0,
                ..Default::default()
            },
        ];
        let decimals = 18;
        let score = votes.iter().map(|v| v.voting_power).sum::<f64>();
        let limit = 2000; // 20 %

        adjust_vote_weights(&mut votes, decimals, score, limit).unwrap();

        assert_eq!(votes[0].voting_power, 200.0);
        assert_eq!(votes[1].voting_power, 200.0);
        assert_eq!(votes[2].voting_power, 200.0);
        assert_eq!(votes[3].voting_power, 200.0);
        assert_eq!(votes[4].voting_power, 83.33333333333333);
        assert_eq!(votes[5].voting_power, 66.66666666666666);
        assert_eq!(votes[6].voting_power, 50.0);
    }
}
