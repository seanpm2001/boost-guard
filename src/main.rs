use axum::routing::{get, post};
use axum::{Extension, Router};
use boost_guard::routes::{handle_create_vouchers, handle_get_rewards, handle_health};
use std::env;
use std::net::SocketAddr;
use std::str::FromStr;
use tokio::net::TcpListener;
extern crate dotenv;

use dotenv::dotenv;

#[tokio::main]
async fn main() {
    dotenv().ok();

    let port: u16 = env::var("PORT")
        .map(|val| val.parse::<u16>().unwrap())
        .unwrap_or(8080);
    let addr = SocketAddr::from(([0, 0, 0, 0], port));
    let listener = TcpListener::bind(addr).await.unwrap();

    axum::serve(listener, app()).await.unwrap();
}

fn app() -> Router {
    dotenv().ok();

    let client = reqwest::Client::new();
    let private_key = env::var("PRIVATE_KEY").expect("PRIVATE_KEY must be set");
    let wallet = ethers::signers::LocalWallet::from_str(&private_key)
        .expect("failed to create a local wallet");
    let state = boost_guard::State { client, wallet };

    Router::new()
        .route("/create-vouchers", post(handle_create_vouchers))
        .route("/get-rewards", post(handle_get_rewards))
        .route(
            "/get-lottery-winners",
            post(boost_guard::routes::handle_get_lottery_winners),
        )
        .route("/health", get(handle_health))
        .layer(Extension(state))
}

#[cfg(test)]
mod tests {
    use axum::body::Body;
    use axum::http;
    use boost_guard::routes::{
        CreateVouchersResponse, GetLotteryWinnerQueryParams, GetLotteryWinnersResponse,
        GetRewardsResponse, QueryParams,
    };
    use http_body_util::BodyExt;
    use tower::ServiceExt;

    // TODO: those test use fixed proposals and voter addresses, but these change from time to time as we delete
    // proposals from the hub... we should probably settle for a fixed proposal and voter address and use those
    #[tokio::test]
    async fn test_create_vouchers() {
        let app = super::app();
        let query = QueryParams {
            proposal_id: "0x5c4d271f77150458cb0265cd5b473dd970bb5fa1fe4e006775c52b94c8e363a1"
                .to_string(),
            voter_address: "0xc83A9e69012312513328992d454290be85e95101".to_string(),
            boosts: vec![("0".to_string(), "1".to_string())],
        };

        let response = app
            .oneshot(
                http::Request::builder()
                    .method(http::Method::POST)
                    .uri("/create-vouchers")
                    .header(http::header::CONTENT_TYPE, mime::APPLICATION_JSON.as_ref())
                    .body(Body::from(serde_json::to_vec(&query).unwrap()))
                    .unwrap(),
            )
            .await
            .unwrap();

        let bytes = response.into_body().collect().await.unwrap().to_bytes();
        let response: Result<Vec<CreateVouchersResponse>, _> = serde_json::from_slice(&bytes);
        if response.is_err() {
            println!("ERROR: {}", String::from_utf8(bytes.to_vec()).unwrap());
            panic!();
        } else {
            println!("OK: {:?}", response.unwrap());
        }
    }

    #[tokio::test]
    async fn test_get_rewards() {
        let app = super::app();
        let query = QueryParams {
            proposal_id: "0x5c4d271f77150458cb0265cd5b473dd970bb5fa1fe4e006775c52b94c8e363a1"
                .to_string(),
            voter_address: "0xc83A9e69012312513328992d454290be85e95101".to_string(),
            boosts: vec![("0".to_string(), "1".to_string())],
        };

        let response = app
            .oneshot(
                http::Request::builder()
                    .method(http::Method::POST)
                    .uri("/get-rewards")
                    .header(http::header::CONTENT_TYPE, mime::APPLICATION_JSON.as_ref())
                    .body(Body::from(serde_json::to_vec(&query).unwrap()))
                    .unwrap(),
            )
            .await
            .unwrap();

        let bytes = response.into_body().collect().await.unwrap().to_bytes();
        let response: Result<Vec<GetRewardsResponse>, _> = serde_json::from_slice(&bytes);
        if response.is_err() {
            println!("ERROR: {}", String::from_utf8(bytes.to_vec()).unwrap());
            panic!();
        } else {
            println!("OK: {:?}", response.unwrap());
        }
    }

    #[tokio::test]
    async fn test_get_lottery_winners() {
        let app = super::app();
        let query = GetLotteryWinnerQueryParams {
            proposal_id: "0xe2f9d5694f6af28b61b357b80752566654b026343dbdf09d36d291b6325dedb3"
                .to_string(),
            boost_id: "12".to_string(),
            chain_id: "11155111".to_string(),
        };

        let response = app
            .oneshot(
                http::Request::builder()
                    .method(http::Method::POST)
                    .uri("/get-lottery-winners")
                    .header(http::header::CONTENT_TYPE, mime::APPLICATION_JSON.as_ref())
                    .body(Body::from(serde_json::to_vec(&query).unwrap()))
                    .unwrap(),
            )
            .await
            .unwrap();

        let bytes = response.into_body().collect().await.unwrap().to_bytes();
        let response: Result<GetLotteryWinnersResponse, _> = serde_json::from_slice(&bytes);
        if response.is_err() {
            println!("ERROR: {}", String::from_utf8(bytes.to_vec()).unwrap());
            panic!("failed test");
        } else {
            println!("OK: {:?}", response.unwrap());
        }
    }

    #[tokio::test]
    async fn test_health_check() {
        let app = super::app();
        let response = app
            .oneshot(
                http::Request::builder()
                    .method(http::Method::GET)
                    .uri("/health")
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();
        assert!(response.status() == http::StatusCode::OK);
    }
}
