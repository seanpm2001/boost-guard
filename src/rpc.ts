import express from 'express';
import { Wallet } from '@ethersproject/wallet';
import { rpcError, rpcSuccess } from './utils';
import { Guard } from './guard';

const privateKey = process.env.GUARD_PK || '';
const wallet = new Wallet(privateKey);
const guard = new Guard(wallet);

const router = express.Router();

router.post('/', async (req, res) => {
  const { id, params } = req.body;
  try {
    const { boostId, recipient } = params;
    const result = await guard.run(boostId, recipient);
    const claim = await guard.claim(result);
    return rpcSuccess(res, claim, id);
  } catch (e) {
    return rpcError(res, 500, e, id);
  }
});

export default router;
