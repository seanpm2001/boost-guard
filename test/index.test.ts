import { Guard } from '../src/guard';
import { Wallet } from '@ethersproject/wallet';

const GUARD_PK = '0x3b6b2589512955bcc0514cc4ee37b546098342b0ed8575f762ea677a71e4d051';

describe('', () => {
  const wallet = new Wallet(GUARD_PK);
  const guard = new Guard(wallet);
  const boostId = '0x1';
  const recipient = '0xeF8305E140ac520225DAf050e2f71d5fBcC543e7';

  it('run strategy', async () => {
    const result = await guard.run(boostId, recipient);
    expect(result).toMatchSnapshot();
  });

  it('issue claim', async () => {
    const claim = await guard.claim({
      boostId,
      recipient,
      amount: 1
    });
    expect(claim).toMatchSnapshot();
  });
});
