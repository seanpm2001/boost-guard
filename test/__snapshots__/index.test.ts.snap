// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` issue claim 1`] = `
Object {
  "address": "0x04c85ed866d6aa29d7888fae80d4e474a9b38b3ecbae00806aea0fdb0314c5ae8c819ed192bc16a44670e496f5d4b8742143caa59578f0b920c530aa3049203cfb",
  "data": Object {
    "domain": Object {
      "chainId": 4,
      "name": "boost",
      "verifyingContract": "0x0000000000000000000000000000000000000000",
      "version": "1",
    },
    "message": Object {
      "amount": 1,
      "boostId": "0x1",
      "recipient": "0xeF8305E140ac520225DAf050e2f71d5fBcC543e7",
    },
    "types": Object {
      "Claim": Array [
        Object {
          "name": "boostId",
          "type": "uint256",
        },
        Object {
          "name": "recipient",
          "type": "address",
        },
        Object {
          "name": "amount",
          "type": "uint256",
        },
      ],
    },
  },
  "sig": "0x9a4c19cc311839a47be20aa76df6fa8c40ebbc6d75cd9d0b6f650dd14cfa4f703b8f3edb1c1723c396c9eef7565530e93542ea5ec4f83f4f320e47bb5ce0dd0e1c",
}
`;

exports[` run strategy 1`] = `
Object {
  "amount": 6,
  "boostId": "0x1",
  "recipient": "0xeF8305E140ac520225DAf050e2f71d5fBcC543e7",
}
`;
